### ISSUE TITLE
Player sub classes

### ISSUE THREAD

**Issue by MaanooAk** [SFTtech/openage#842]
**Player**: Contains information and operations of game Player

- `Player` 
// hint: they should have a different data structure or different operations
  - `Human Player`
  - `AI Player`
  - `Idle Player` (for Gaia or multiple independent Gaia**s**) (or even `Gaia Player`)
  - `Dummy Player` (usually provided in many modding environments, eg. used as a bank)

**Controller**: Takes input from a source (eg. gui, AI) and transforms them into game actions

- `Controller` 
// hint: they should have a different input source or different generated game actions
  - `Human Controller` (gui actions to actions)
  - `AI Controller` (AI algorithms to actions)
    - `Scripted Controller` (Combines fixed actions and actions from AI algorithms)
  - `Empty Controller` (do nothing)

#### Old post
> At some point the class Player will become the base class of something like this
> 
> - `Player`
>   - `Human Player`
>   - `AI Player`
>     - `Scripted Player` (for campaign like stuff)
>   - `Idle Player` (for Gaia or multiple independent Gaia**s**) 
>   - `Dummy Player` (usually provided in many modding environments)
> 
> Any suggestions on the topic? different tree? other sub classes?

**Comment by TheJJ** [comment]
Hmm, not bad. Up until now I thought that it should be designed in such a way that there's only one `Player` and that all actions of other (e.g. autonomous) Players are done by having a dispatching frontent that delivers things to the same player, but can receive input from e.g. real players, or AI.
What do you think?

**Comment by MaanooAk** [comment]
At the moment Player is mostly used as a "data type" with a minimal functionality, I believe that more things should go inside (eg. what I have been doing in my last prs: the score keeping system, the population system). In that case:
1. If you have only one class at some point you will have code that looks like "if human ... else if ai ... else if gaia ..." eg. in cases where you need to cheat (eg. campaign player no need to follow building dependecies) or other things
2. For optimization purposes eg. no need for gaia to keep score or population information
3. Changes of the core elements eg. in gaia we may have a separate specialized data structure for only storing the trees
4. More good things I haven't thought 

Just to clarify, this is to work in combination with the dispatching idea (= the AI Player will not contain the actual AI)

**Comment by TheJJ** [comment]
Indeed, and the base `Controller` owns the `Player` object. The other controllers then extend the API so that e.g. the `AIController` has the functions needed to control the `Player` with things from the AI.
The difference between our ideas is pretty minimal:

```
yours:
  Player
     `- AIPlayer
     `- Humanplayer

mine:
  Controller  ------------> Player
     `- AIController
     `- HumanController
```

I'd just add the layer in front of the player, you below it.
Which one is better I can't really tell, I think it might be better to have a unified player which operates completely independent of its controller. And the controller needs specializations. The outcome is the same though, your `AIPlayer` could just have the controlling functions my `HumanController` would provide.
I think both solutions do pretty much the same thing. Mine is more abstracted, yours is simpler.

**Comment by MaanooAk** [comment]
> Just to clarify, this is to work in combination with the dispatching idea (= the AI Player will not contain the actual AI)

I meant this:
```
Controller                       ---->    Player
Human Controller : Controller    ---->    Human Player : Player
AI Controller : Controller       ---->    AI Player : Player
Scripted Controller : AI Contr.  ---->    AI Player : Player
Empty Controller : Controller    ---->    Idle Player : Player
Empty Controller : Controller    ---->    Dummy Player : Player
```

(I updated the the post)

**Comment by TheJJ** [comment]
That is pretty much what I thought of, except that I didn't anticipate the need for specialized players. But we will see, you'll do it in a good way surely :)

**Comment by TheJJ** [comment]
If there's anything left to discuss we can surely reopen :)

**Comment by TheJJ** [comment]
Hm, we I think it'll get buried if we don't leave it open until we implement it or obsolete it.
